x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/plugin/TransportEqlSearchAction.java

```diff
@@ -122,8 +134,52 @@ public static void operation(PlanExecutor planExecutor, EqlSearchTask task, EqlS
 
         EqlConfiguration cfg = new EqlConfiguration(request.indices(), zoneId, username, clusterName, filter, timeout,
                 request.indicesOptions(), request.fetchSize(), clientId, new TaskId(nodeId, task.getId()), task);
-        planExecutor.eql(cfg, request.query(), params, wrap(r -> listener.onResponse(createResponse(r, task.getExecutionId())),
-            listener::onFailure));
+        Holder<Boolean> retrySecondTime = new Holder<Boolean>(false);
+        planExecutor.eql(cfg, request.query(), params, wrap(r -> listener.onResponse(createResponse(r, task.getExecutionId())), e -> {
+            // the search request will likely run on nodes with different versions of ES
```

Comment:
I think I put "likely" there because the actual search might run on a sub-set of shards than the one at the start of the search process. There is an initial phase in the search process that's called `canMatch` that is really quick and can evaluate if a request on a shard should actually be executed on the shard or not, based on a timestamp `range` query. If I am not mistaken, this `canMatch` phase is also part of this different versions mismatch.
The `canMatch` can quickly "discard" some shards and the actual request truly runs on a sub-set of them. If you want to read more about this, I think this is the initial PR adding this - https://github.com/elastic/elasticsearch/pull/25658. And some other good resources here: https://stackoverflow.com/a/64693782/3498062

Reply:
Thanks. I think what threw me off is the future tense, in a failure handler :-).
I guess "the search request _likely **ran**_ on nodes with different versions of ES" might be clearer. Anyways, not super relevant.

Reply:
Changed